### PR TITLE
⚠️ evaluating sonarqube

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Just exploring how sonarqube looks like and what are the quality signals in place

### CODE REVIEW

**Comment by rhass** [comment]
@v1v 

When running in Buildkite, you will need to set these values in your build:

https://github.com/elastic/sonarqube/blob/74e547e6a05b921416d4b3f19571dc8a42dae88a/buildkite-scanner/scan-source-code.sh#L8-L11

Without this, the PR scanning wont work and it will do branch analysis rather than PR level.

**Comment by v1v** [comment]
Thanks @rhass , I'm using GH actions, it seems Maven projects need to use the maven sonar plugin instead, so I'm giving a go and see how it goes

**Comment by rhass** [comment]
@v1v The maven plugin should work well. :+1:  

Once you have this all hooked up, I think you may need set those `-Dsonar.xxxx` parameters for your maven build to tell sonar to use PR scanning rather than "branch" scanning. This will do a few things, including automatically add the sonarqube results to the PR, and instruct Sonar only to pay attention to the changes in a PR rather than the entire repo so that devs don't run into failures unrelated to the content of their own changes.

Feel free to ping me on slack too if you have any questions! :)

**Comment by v1v** [comment]
> Feel free to ping me on slack too if you have any questions! :)

Definitely, Thanks so much for being on top and those hints! 🙏

**Comment by v1v** [comment]
I'm seeing errors with

```
2023-03-23T20:46:54.4540428Z [ERROR] Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.9.1.2184:sonar (default-cli) on project apm-agent-parent: Error creating bean with name 'org.sonarsource.scanner.api.internal.IsolatedClassloader@26225c5e-org.sonar.scanner.bootstrap.ScannerPluginRepository': Initialization of bean failed; nested exception is java.lang.NoClassDefFoundError: java/sql/Time: java.sql.Time -> [Help 1]
2023-03-23T20:46:54.4541361Z [ERROR] 
```

and also locally, so I enabled `-e` and got

```
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.9.1.2184:sonar (default-cli) on project apm-agent-parent: Error creating bean with name 'org.sonarsource.scanner.api.internal.IsolatedClassloader@2c6b8ac8-org.sonar.scanner.bootstrap.ScannerPluginRepository': Initialization of bean failed; nested exception is java.lang.NoClassDefFoundError: java/sql/Time: java.sql.Time -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.9.1.2184:sonar (default-cli) on project apm-agent-parent: Error creating bean with name 'org.sonarsource.scanner.api.internal.IsolatedClassloader@2c6b8ac8-org.sonar.scanner.bootstrap.ScannerPluginRepository': Initialization of bean failed; nested exception is java.lang.NoClassDefFoundError: java/sql/Time
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:306)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:211)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:157)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:121)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:127)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:47)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:156)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:72)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error creating bean with name 'org.sonarsource.scanner.api.internal.IsolatedClassloader@2c6b8ac8-org.sonar.scanner.bootstrap.ScannerPluginRepository': Initialization of bean failed; nested exception is java.lang.NoClassDefFoundError: java/sql/Time
    at org.sonarsource.scanner.maven.bootstrap.ScannerBootstrapper.execute (ScannerBootstrapper.java:67)

```

my local env:

```
Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0)
Maven home: /Users/vmartinez/.m2/wrapper/dists/apache-maven-3.8.5-bin/5i5jha092a3i37g0paqnfr15e0/apache-maven-3.8.5
Java version: 17.0.5, vendor: Eclipse Adoptium, runtime: /Users/vmartinez/.sdkman/candidates/java/17.0.5-tem
Default locale: en_GB, platform encoding: UTF-8
OS name: "mac os x", version: "13.2.1", arch: "x86_64", family: "mac"
```

and the one in the CI

```
2023-03-23T20:08:33.8016198Z Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0)
2023-03-23T20:08:33.8017441Z Maven home: /home/runner/.m2/wrapper/dists/apache-maven-3.8.5-bin/5i5jha092a3i37g0paqnfr15e0/apache-maven-3.8.5
2023-03-23T20:08:33.8018090Z Java version: 17.0.6, vendor: Eclipse Adoptium, runtime: /usr/lib/jvm/temurin-17-jdk-amd64
2023-03-23T20:08:33.8018675Z Default locale: en, platform encoding: UTF-8
2023-03-23T20:08:33.8019188Z OS name: "linux", version: "5.15.0-1034-azure", arch: "amd64", family: "unix"
```

**Comment by rhass** [comment]
Hrmm.... It's possible it doesn't have the permissions to download project config/plugin data from the server. Try creating `Project Analysis` token rather than a `User` token. I am not sure if the user tokens are scoped to certain APIs or not, but I have tried using `Project Analysis` and `Global Analysis` tokens when running scans so far.

**Comment by rhass** [comment]
There docs say this:
> User tokens
> These tokens can be used to run analyses and to invoke web services, based on the token author's permissions.

So my theory doesn't seem likely. :thinking:

**Comment by v1v** [comment]
Good catch let me try that.

OTOH, somehow I was feeling it could be related to how I configured the local environment variables rather than using the maven properties. https://github.com/elastic/apm-agent-java/pull/3072/commits/42e0dd51ce6da87dab9bc79bffe96511689c6e73

I'll give another swing after this one with `Project Analysis`, when i created the token I was not sure which one to use, so I took the `User` one instead.

Thanks again!

File: .github/workflows/sonarqube.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      - uses: hashicorp/vault-action@v2.4.2
+        with:
+          url: ${{ secrets.VAULT_ADDR }}
+          method: approle
+          roleId: ${{ secrets.VAULT_ROLE_ID }}
+          secretId: ${{ secrets.VAULT_SECRET_ID }}
+          secrets: |
+            secret/observability-team/ci/sonarqube/apm-agent-java sonar-token | SONAR_TOKEN ;
+            secret/observability-team/ci/sonarqube/apm-agent-java sonar-host-url | SONAR_HOST_URL ;
+            secret/observability-team/ci/sonarqube/apm-agent-java project-key | SONAR_PROJECT_KEY
```
Comment (rhass) [inline .github/workflows/sonarqube.yml:40]: Do these values get scoped to the other steps?

**Comment by rhass** [comment]
I checked out this branch and ran `./mvnw sonar:sonar -Dsonar.projectKey="${SONAR_PROJECT}" -Dsonar.login="${SONAR_LOGIN}" -Dsonar.host.url="${SONAR_HOST}"` and got back pretty much the same error. It looks like this is a classpath issue, but I am not sure why it is happening. 

```
Caused by: java.lang.NoClassDefFoundError: java/sql/Time
    at com.google.gson.Gson.<init> (Gson.java:265)
    at com.google.gson.Gson.<init> (Gson.java:186)
    at org.sonar.scanner.bootstrap.ScannerPluginInstaller.listInstalledPlugins (ScannerPluginInstaller.java:105)
```

This may be a red herring, but I found someone had a similar issue with this class which was apparently caused by some classpath shortener. 

https://stackoverflow.com/questions/57222110/gson-and-java-sql-in-java-11

:shrug: 

Usually the sonar maven plugin is pretty straight forward, but I think classpath is a good start to look. 

Depending on how your unit and ITs are executed, you could also use the current coverage GitHub actions steps you have, and add the `jacoco:report-aggregate` goal to generate a single report file, then save it as an artifact which you can fetch in a separate sonar step that uses the SonarQube action to ingest the reports. The issue here is that it becomes a stateful action that needs to wait for something else to complete to run.

**Comment by rhass** [comment]
Hello! Just wanted to check in and see if there is anything we can do to help you out.

**Comment by v1v** [comment]
Hey @rhass , I've been  on PTO, I plan to revisit this next week, I'll ping you with my findings 🙏

**Comment by v1v** [comment]
run docs-build

**Pull request closed without merging**